### DISCUSSION TITLE
Race condition when an action creates and then modifies a repository via the API

### DISCUSSION THREAD

**Discussion by chrisnewark** [community/community#26333] [Actions]
<p>I have a job that creates a new repository (using a Github App token) and then creates a branch protection rule immediately after.</p>
<pre><code class="lang-auto">create-repo:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Create repo
        run: gh repo create $REPO_PATH --private --template $TEMPLATE_PATH
      - name: Disable repo features
        run: gh repo edit $REPO_PATH --enable-auto-merge=false --enable-issues=false --enable-projects=false --enable-wiki=false
      - name: Set branch protection
        run: gh api repos/$REPO_PATH/branches/main/protection --input ${{ github.workspace }}/.github/workflows/api-payloads/branch-protection.json --method PUT
</code></pre>
<p>I am finding that some (maybe 1 in 4) executions of the <code>Set branch protection</code> fail stating the main branch is not found (404). I seem to have resolved the problem by introducing a 5 second sleep task before <code>Set branch protection</code>.</p>
<p>Is this definitely caused by some sort of race condition between the repo being created and the branch being available for modification and if it is, is there a more satisfactory solution then a sleep?</p>

**Comment by otter-computer** [comment, marked as answer]
<p>Hey <a class="mention" href="/chrisnewark">@chrisnewark</a>, welcome to the GitHub Community! :sparkles:</p>
<p>I’m not really sure there <em>is</em> going to be a more elegant solution that just adding a sleep here to be honest. It sounds like the background job that creates the repository just isn’t completing in time before you try to edit the protected branch.</p>
<p>You could use the API to try to <a href="https://docs.github.com/en/rest/branches/branches#get-a-branch">get the branch</a> and, if it doesn’t exist, sleep then try again, but that does seem like it’s just adding more complexity around the same thing. Perhaps you would want to do this if you run this workflow often and those sleep seconds add up in your billing time?</p>
<p>When you do run this workflow are you seeing the <code>201</code> status code response  from the <code>gh repo create</code> command?</p>

**Comment by chrisnewark** [comment]
<p>Hi <a class="mention" href="/thomasshaped">@ThomasShaped</a>, thanks for your reply!</p>
<p>I’m unable to see the response code when using the gh cli as the standard output is just the new repo url. Unsure if the cli has any sort of debug flag I can use. I tested this with a direct request to the REST API and I did indeed get a 201.</p>
<p>This workflow isn’t run very often so the 5 second sleep will be an adequate solution in this case. Thanks!</p>
